src/db_impl.cc

```diff
@@ -175,114 +175,144 @@ Status TitanDBImpl::ValidateOptions(
-    base_descs.emplace_back(desc.name, desc.options);
+  if (!s.ok()) {
+    return s;
   }
-  std::map<uint32_t, TitanCFOptions> column_families;
-
-  // Opens the base DB first to collect the column families information
-  //
-  // Disable compaction at this point because we haven't add table properties
-  // collector. A compaction can generate a SST file without blob size table
-  // property. A later compaction after Titan DB open can cause crash because
-  // OnCompactionCompleted use table property to discover blob files generated
-  // by the compaction, and get confused by missing property.
-  //
-  // We also avoid flush here because we haven't replaced the table factory
-  // yet, but rocksdb may still flush if memtable is full. This is fine though,
-  // since values in memtable are raw values.
-  for (auto& desc : init_descs) {
-    desc.options.disable_auto_compactions = true;
-  }
-  db_options_.avoid_flush_during_recovery = true;
-  // Add EventListener to collect statistics for GC
-  db_options_.listeners.emplace_back(std::make_shared<BaseDbListener>(this));
   // Note that info log is initialized after `CreateLoggerFromOptions`,
   // so new `BlobFileSet` here but not in constructor is to get a proper info
   // log.
   blob_file_set_.reset(new BlobFileSet(db_options_, stats_.get()));
-
-  s = DB::Open(db_options_, dbname_, init_descs, handles, &db_);
-  if (s.ok()) {
-    for (size_t i = 0; i < descs.size(); i++) {
-      auto handle = (*handles)[i];
-      uint32_t cf_id = handle->GetID();
-      std::string cf_name = handle->GetName();
-      column_families.emplace(cf_id, descs[i].options);
-      db_->DestroyColumnFamilyHandle(handle);
-      // Replaces the provided table factory with TitanTableFactory.
-      // While we need to preserve original table_factory for GetOptions.
-      auto& base_table_factory = base_descs[i].options.table_factory;
-      assert(base_table_factory != nullptr);
-      auto titan_table_factory = std::make_shared<TitanTableFactory>(
-          db_options_, descs[i].options, blob_manager_, &mutex_,
-          blob_file_set_.get(), stats_.get());
-      cf_info_.emplace(cf_id,
-                       TitanColumnFamilyInfo(
-                           {cf_name, ImmutableTitanCFOptions(descs[i].options),
-                            MutableTitanCFOptions(descs[i].options),
-                            base_table_factory, titan_table_factory}));
-      base_descs[i].options.table_factory = titan_table_factory;
-      // Add TableProperties for collecting statistics GC
-      base_descs[i].options.table_properties_collector_factories.emplace_back(
-          std::make_shared<BlobFileSizeCollectorFactory>());
-    }
-    handles->clear();
-    s = db_->Close();
-    delete db_;
-    db_ = nullptr;
-  }
-  if (!s.ok()) return s;
-
-  if (stats_.get()) {
-    stats_->Initialize(column_families);
+  // Setup options.
+  db_options_.listeners.emplace_back(std::make_shared<BaseDbListener>(this));
```

Comment:
seems this should set before `blob_file_set.reset` too.
To avoid other similar issues, maybe we should pass `db_option` when `BlobFileSet::Open`

Reply:
why? does blob_file_set use the listeners?

Reply:
Oh, I misread the code.